Basic `ImmersedBoundaryGrid` + `GridFittedBoundary` implementation

glwagner opened the pull request:
This PR implements a primitive immersed boundary scheme that is "fitted" to the grid.

The scheme is specified with a function of `(x, y, z)` that is true when "solid" and false when "fluid". (A better name for the model kwarg than `immersed_boundary` might be `immersed_solid`.) It only supports no flux boundary conditions, but it should be "relatively" straightforward to support user specified flux boundary conditions.

The scheme has the following ingredients:

* Masking of solid cells to 0 for both velocities and tracers.
* Elides diffusive fluxes across solid boundaries with an explicit `ifelse` embedded in diffusive flux and viscous flux functions.

Because of the second point the scheme is "invasive" in that it requires the immersed boundary to be inserted into the tendency calculation. On the upside, it is likely cheap since it doesn't require any extra loops (except for the masking).

EDIT: the implementation in this PR is _very_ invasive in that it introduces a grid wrapper that implements the immersed boundary functionality. Grid wrappers make the implementation of more sophisticated immersed boundary methods such as partial cell and shaved cells much easier.

To support user-specified fluxes, we'd have to distinguish between a `solid_node` and a `fluid_solid_boundary_node`, and drop in the user-specified flux when appropriate.

There's a primitive validation test in `validation/immersed_boundary/immersed_hydrostatic_diffusion.jl` that produces this plot:

<img width="587" alt="image" src="https://user-images.githubusercontent.com/15271942/117931760-0dc40000-b2ac-11eb-8978-2c376d706948.png">

I'm not sure how the immersed boundary will interact with the free surface, so we'd have to test that.

In order to implement this scheme in `IncompressibleModel`, we'll have to get rid of the existing implementation. It's also unclear how accurate either this or the previous approximation is without modification of the pressure solver algorithm.

I think to merge this we need 1) unit tests and 2) a validation tests that exercises the momentum equations. Supporting user-specified fluxes can come in a subsequent PR.

Some of the pieces to support immersed boundaries with VerticallyImplicitTimeDiscretization are in place but it's not 100% supported yet.

cc @hdrake @ali-ramadhan

glwagner commented:
This script:

```julia
using Oceananigans, Plots

grid = RegularRectilinearGrid(size=(128, 128), x=(-5, 5), z=(0, 5), topology=(Periodic, Flat, Bounded))

# Gaussian bump
bump(x, y, z) = z < exp(-x^2)

# Tidal forcing
tidal_forcing(x, y, z, t) = 1e-2 * cos(t)

model = HydrostaticFreeSurfaceModel(grid = grid,
                                    momentum_advection = UpwindBiasedThirdOrder(),
                                    closure = IsotropicDiffusivity(ν = 1e-6, κ = 1e-6),
                                    tracers = :b, 
                                    buoyancy = BuoyancyTracer(),
                                    immersed_boundary = bump,
                                    forcing = (u = tidal_forcing,))
    
# Linear stratification with N² = 1
set!(model, b = (x, y, z) -> z)
    
simulation = Simulation(model, Δt = 2e-3, stop_iteration=10000)
run!(simulation)

xu, yu, zu = nodes(model.velocities.u)
u = interior(model.velocities.u)[:, 1, :]
u_plot = heatmap(xu, zu, u', title="u velocity")

xb, yb, zb = nodes(model.tracers.b)
b = interior(model.tracers.b)[:, 1, :]
b_plot = heatmap(xb, zb, b', title="buoyancy")

ub_plot = plot(u_plot, b_plot, layout=(2, 1)) 
display(ub_plot)
```

produces

<img width="804" alt="image" src="https://user-images.githubusercontent.com/15271942/117941274-68faf000-b2b6-11eb-8eeb-e82abae50b82.png">

The script is at `validation/immersed_boundary/internal_tide.jl`.

Seems unfortunate that the time-step needs to be so short but perhaps that's right?

More work needed.

glwagner commented:
Hmm, I think my parameters are silly because they aren't hydrostatic, I don't think. Thinking rather than typing is needed now.

francispoulin commented:
That is a very exciting example you have there @glwagner !  Great start.

I have a colleague who has done something similar in a bunch of papers.  Here is [one](https://www.sciencedirect.com/science/article/pii/S0278434307000192?casa_token=10M4stPzKB4AAAAA:9_zlffVVdKfAiOdq7MdGGIiha3L6CbbX3AmBqen60EJb9FaP9Nj8I9I4VecUqmxO0B_ls_Tq_A) that *might* be of interest.  Figure 1 certainly produces some very nice patterns.

ali-ramadhan commented on src/ImmersedBoundaries/mask_immersed_field.jl:
```diff
@@ -0,0 +1,15 @@
+    return launch!(arch, field.grid, :xyz, _mask_immersed_field!, field, loc, ib; dependencies = device_event(arch))
+end
+
+@kernel function _mask_immersed_field!(field, loc, ib)
+    i, j, k = @index(Global, NTuple)
+    @inbounds field[i, j, k] = ifelse(solid_node(loc..., i, j, k, ib.grid, ib), 0, field[i, j, k])
+end
```
I guess this is the only part that needs to be modified for `GridFittedImmersedBoundary` to work with `IncompressibleModel`?

Since modifying the velocities would cause the velocity field to become divergent, you'd want to relax the velocity instead like in https://github.com/CliMA/Oceananigans.jl/blob/57ac2d2cf16f08218037376c7af832ec457940e7/src/Models/IncompressibleModels/correct_incompressible_immersed_tendencies.jl#L43-L65

glwagner replied:
For `IncompressibleModel` you make a choice: either 1) the boundary pressure gradient / boundary conditions are wrong, or 2) the velocity field is divergent.

I don't think relaxation solves the issue --- but perhaps your hypothesis is that is mitigates the problem? I'm not sure. Either way, tracer is not conserved.

The tracer conservation issue can be _solved_ by eliding advective fluxes in addition to diffusive fluxes on the boundary.

The larger issue can be solved by modifying the pressure solver. Perhaps the `PreconditionedConjugateGradientSolver` to "polish off" the pressure field obtained naively without considering boundary conditions, or we can design an iterative scheme with the FFT based pressure solver.

glwagner replied:
Ah I misunderstood this comment originally. I believe the quoted function in `correct_incompressible_immersed_tendencies.jl` is _identical_ to `mask_immersed_fields.jl`. I believe the correction just calculates `Gⁿ` such that the RK update masks the velocity to 0. I believe `mask_immersed_fields!` is just a more straightforward way of achieving the same thing. So the current implementation doesn't implement relaxation.

glwagner replied:
So I guess with the functionality provided by this PR you could achieve identical functionality as previously (except with no flux boundary conditions) by applying `mask_immersed_fields!` to the predictor velocity rather than in `update_state!`.

In that case tracer leaks across the boundary. But this can be fixed by eliding advective fluxes in addition to diffusive fluxes.

It'd be worth investigating whether masking of all fields in `update_state!` _in addition to_ masking of the predictor velocities prior to the pressure solve results in a non-divergent velocity field that satisfies boundary conditions.

ali-ramadhan reviewed (approved):
This looks great! I can adapt `GridFittedImmersedBoundary` for `IncompressibleModel` in a separate PR if that makes sense.

cc @gianlisi @wanyingkang who would be interested in this too.

ali-ramadhan reviewed (commented):
Woops prematurely approved by mistake. Taking it back.

ali-ramadhan reviewed (changes requested):
Hmmm gotta request changes to take back an approval?

glwagner commented:
> I can adapt `GridFittedImmersedBoundary` for `IncompressibleModel` in a separate PR if that makes sense?

I think a separate PR is best. It's really a research PR since we don't have a recipe so the validation process will probably be more involved. It's not clear what algorithm we should choose for `IncompressibleModel`.

glwagner commented:
> Here is [one](https://www.sciencedirect.com/science/article/pii/S0278434307000192?casa_token=10M4stPzKB4AAAAA:9_zlffVVdKfAiOdq7MdGGIiha3L6CbbX3AmBqen60EJb9FaP9Nj8I9I4VecUqmxO0B_ls_Tq_A) that _might_ be of interest. Figure 1 certainly produces some very nice patterns.

Thanks @francispoulin, that's helpful!

navidcy commented:
> Hmm, I think my parameters are silly because they aren't hydrostatic, I don't think. Thinking rather than typing is needed now.

Callum would know. I can ask him about parameters, dt, etc that should work OK.

navidcy commented:
Here are some parameters from Callum:

`Δx = 100 m`, `Nx = 1024`
`Δz = 1.5 m`, `Nz = 3000`
`Δt = 15 s`
`N² = 2e-6 s⁻¹`
`depth = - 3000 m + 100 m * exp(-(x / 3km)²)`
`f =1e-4 s⁻¹`
`tidal frequency = 1.4e-4 s⁻¹`
`tidal velocity `0.015 m s⁻¹`

horizontal viscosity/diffusivity
`νₕ = κₕ = 1e-2 m² s⁻¹`
vertical viscosity/diffusivity
`νᵥ = κᵥ = 1e-4 m² s⁻¹`

glwagner commented:
> > Hmm, I think my parameters are silly because they aren't hydrostatic, I don't think. Thinking rather than typing is needed now.
> 
> 
> 
> Callum would know. I can ask him about parameters, dt, etc that should work OK. 

I believe to be hydrostatic  we need a small aspect ratio and forcing far from N.

navidcy commented:
> > > Hmm, I think my parameters are silly because they aren't hydrostatic, I don't think. Thinking rather than typing is needed now.
> > 
> > 
> > Callum would know. I can ask him about parameters, dt, etc that should work OK.
> 
> I believe to be hydrostatic we need a small aspect ratio and forcing far from N.

Well Callum actually ran the simulation with the parameters above on MITgcm in hydrostatic mode.

glwagner commented:
> > > > Hmm, I think my parameters are silly because they aren't hydrostatic, I don't think. Thinking rather than typing is needed now.
> > > 
> > > 
> > > Callum would know. I can ask him about parameters, dt, etc that should work OK.
> > 
> > 
> > I believe to be hydrostatic we need a small aspect ratio and forcing far from N.
> 
> Well Callum actually ran the simulation with the parameters above on MITgcm in hydrostatic mode.
> 
> I put it there because it might give a hint for something going wrong if, eg, you meed 100 times less `Δt` to get a stable simulation with everything else the same.

Ah I missed the `Δt` --- is that maybe limited by advection / CFL? It's certainly very far from 1 / 12.421 hours...

For my above example I realized I was also using `ExplicitFreeSurface` which I think was likely providing the time step restriction... I think for this validation case I will target a linear regime in which the time step is limited by the highest-frequency internal waves.

glwagner commented:
I refactored the interface significantly. Now instead of using a model property `immersed_boundary`, I've introduced a new grid "wrapper" called `ImmersedBoundaryGrid`. This new grid wrapper replicates the behavior of the underlying or "primary" grid, except where functions are overridden. This allows functions like `diffusive_flux_x` to be extended without touching model time-stepping code.

The main motivation for this change is to make it easy to implement more sophisticated immersed boundary methods such as partial cells and shaved cells. In those methods, the presence of the immersed boundary is felt at a fundamental level during the calculation of grid areas and volumes.

The `internal_tide.jl` validation experiment has also been refactored and now reads:

```julia
using Oceananigans
using Oceananigans.ImmersedBoundaries: ImmersedBoundaryGrid, GridFittedBoundary
using Plots

grid = RegularRectilinearGrid(size=(256, 256), x=(-10, 10), z=(0, 5), topology=(Periodic, Flat, Bounded))

# Gaussian bump of width "1"
bump(x, y, z) = z < exp(-x^2)

grid_with_bump = ImmersedBoundaryGrid(grid, GridFittedBoundary(bump))

# Tidal forcing
tidal_forcing(x, y, z, t) = 1e-4 * cos(t)

model = HydrostaticFreeSurfaceModel(grid = grid_with_bump,
                                    momentum_advection = CenteredSecondOrder(),
                                    free_surface = ExplicitFreeSurface(gravitational_acceleration=4),
                                    closure = IsotropicDiffusivity(ν=1e-4, κ=1e-4),
                                    tracers = :b, 
                                    buoyancy = BuoyancyTracer(),
                                    coriolis = FPlane(f=sqrt(0.5)),
                                    forcing = (u = tidal_forcing,))

# Linear stratification
set!(model, b = (x, y, z) -> 2 * z)

progress(s) = @info "Time = $(s.model.clock.time), max|w|: $(maximum(abs, model.velocities.w))"
                  
simulation = Simulation(model, Δt = 1e-3, stop_time = 10, progress = progress, iteration_interval = 10) 
                            
run!(simulation)

xu, yu, zu = nodes(model.velocities.u)
u = interior(model.velocities.u)[:, 1, :]

umax = maximum(abs, u)
ulim = umax / 2 
ulevels = vcat(-[umax], range(-ulim, stop=ulim, length=30), [umax]) 

xu2 = reshape(xu, grid.Nx, 1)
zu2 = reshape(zu, 1, grid.Nz)
u[bump.(xu2, 0, zu2)] .= NaN 

u_plot = heatmap(xu, zu, u'; title = "x velocity", color = :balance, clims = (-ulim, ulim))

display(u_plot)
```

Note that the small time-step is near the gravity wave CFL for an explicit free surface on the specified grid. Somewhat larger time-steps are possible with an `ImplicitFreeSurface` but it seems the overall integration is slower that way.

glwagner commented:
<3 GPU:

https://user-images.githubusercontent.com/15271942/118384512-fbe7a300-b5b2-11eb-9096-840b3a4a5265.mp4

ali-ramadhan commented on src/Grids/Grids.jl:
```diff
@@ -80,19 +80,27 @@ Abstract supertype for grids with elements of type `FT` and topology `{TX, TY, T
 abstract type AbstractGrid{FT, TX, TY, TZ} end
 
+"""
+    AbstractPrimaryGrid{FT, TX, TY, TZ}
+
+Abstract supertype for "primary" grids (as opposed to grids with immersed boundaries)
+with elements of type `FT` and topology `{TX, TY, TZ}`.
```
Out of curiosity is _primary grid_ a common term in the immersed boundary literature?

glwagner replied:
I don't think so. Another term might be "underlying grid". Not sure what's best...

ali-ramadhan replied:
If it's just an internal abstract type then what do you think of `AbstractUnimmersedGrid`? It's is a little verbose but might be clearer?

glwagner replied:
I guess `Primary` or `Underlying` generalizes to other grid wrapper types, but `Unimmersed` would have to be refactored if we had another wrapper type, possibly creating more work for us (I'm not sure how generalized vertical coordinates will be implemented, but it might be via this method).

The other issue is that we have an immersed boundary implementation which would then use `AbstractUnimmersedGrid`, which might be a little confusing, but I'm not sure.

ali-ramadhan replied:
Hmmm so I guess `AbstractUnimmersedGrid` makes sense for this PR but maybe not for future work.

I find `Primary` and `Underlying` kinda vague but the name is explained in the docstring and it shouldn't be hard to change if we come up with a better name.

ali-ramadhan commented on src/ImmersedBoundaries/ImmersedBoundaries.jl:
```diff
@@ -0,0 +1,82 @@
+using Oceananigans.Utils
+using Oceananigans.TurbulenceClosures: AbstractTurbulenceClosure, time_discretization
+
+import Oceananigans.TurbulenceClosures:
+    viscous_flux_ux,
+    viscous_flux_uy,
+    viscous_flux_uz,
+    viscous_flux_vx,
+    viscous_flux_vy,
+    viscous_flux_vz,
+    viscous_flux_wx,
+    viscous_flux_wy,
+    viscous_flux_wz,
+    diffusive_flux_x,
+    diffusive_flux_y,
+    diffusive_flux_z,
+    κᶠᶜᶜ,
+    κᶜᶠᶜ,
+    κᶜᶜᶠ,
+    νᶜᶜᶜ,
+    νᶠᶠᶜ,
+    νᶜᶠᶠ,
+    νᶠᶜᶠ
```
Should `Oceananigans.TurbulenceClosures` be exporting these since it's an internal module now?

glwagner replied:
It could, but that wouldn't help here because we still need explicit `import` to extend these functions...

ali-ramadhan commented on src/ImmersedBoundaries/ImmersedBoundaries.jl:
```diff
@@ -0,0 +1,82 @@
+include("grid_fitted_immersed_boundary.jl")
+include("mask_immersed_field.jl")
+
+#=
+#####
+##### Diffusivities (for VerticallyImplicitTimeDiscretization)
+#####
+
+for (locate_coeff, loc) in ((:κᶠᶜᶜ, (f, c, c)),
+                            (:κᶜᶠᶜ, (c, f, c)),
+                            (:κᶜᶜᶠ, (c, c, f)),
+                            (:νᶜᶜᶜ, (c, c, c)),
+                            (:νᶠᶠᶜ, (f, f, c)),
+                            (:νᶠᶜᶠ, (f, c, f)),
+                            (:νᶜᶠᶠ, (c, f, f)))
+
+    @eval begin
+        @inline $locate_coeff(i, j, k, grid, ib::NoImmersedBoundary, coeff) = $locate_coeff(i, j, k, grid, coeff)
+
+        @inline $locate_coeff(i, j, k, grid::AG{FT}, ib::GFIB, coeff) where FT =
+            ifelse(solid_cell(loc..., i, j, k, grid, ib), $locate_coeff(i, j, k, grid, coeff), zero(FT))
+    end
+end
+=#
```
Remove comments?

glwagner replied:
Hmmmmmmmm actually I should just make this work I think.

ali-ramadhan commented:
I restarted the Buildkite build since the cpu distributed tests failed due to some random ELF error then the build finished and cleaned up the Julia depot so it was too late to restart the cpu distributed tests.

ali-ramadhan commented:
Does `ImmersedBoundaryGrid` + `GridFittedBoundary` only work with `CenteredSecondOrder` advection for now? If so, is it worth adding a warning in case users accidently use higher-order advection schemes with

glwagner commented:
> Does `ImmersedBoundaryGrid` + `GridFittedBoundary` only work with `CenteredSecondOrder` advection for now? If so, is it worth adding a warning in case users accidently use higher-order advection schemes with immersed boundaries?

Unclear --- we need to validate that. You might be right that we need to short-circuit the advective fluxes to generalize to higher-order advection schemes.

francispoulin commented:
Hello @glwagner , we are interested in starting to play with this example to look at flow over a seamount.  What advice do you have before we start looking into this code in detail?

Is there a reason why this is not ready to be merged with master?

glwagner commented:
This code is indeed merged!

francispoulin commented:
Great and thanks!

glwagner commented:
It was merged about a month ago. What functionality are you trying to use?

francispoulin commented:
With my student, I wanted to look at flow over a tall seamount and icemount, like many people have done.  That is a first step, but then I want her to look at sigma-coordinates, as we discussed.  One goal is to compare the two and to learn more about their respective strengths and weaknesses.

I have ran the example you created in this PR on my local computer and it's been 3 hours and it's still going.  Is that to be expected?

glwagner commented:
It looks like the example runs 10,000 iterations so I would expect that to take a while, yes.

francispoulin commented:
Thanks @glwagner .  15 hours and stilll running!

glwagner commented:
I'm pretty sure I ran it on the GPU.

francispoulin commented:
That was a smart idea.  I hope to be more clever next time.

I'm almost 24 hours in so can't stop now

glwagner commented:
Are you running the pasted example or the code in `validation`?

francispoulin commented:
the internal_tide.jl code in validation.  It seemed like the best place to start, even though I know it's just's going to produce the nice animation that you posted in this PR.

glwagner commented:
It's actually a bit different than the example code in the first post. There are some oddities in there that might deserve tweaking. If you improve the validation definitely open a PR!

francispoulin commented:
Ah, I foolishly assumed it was running in CPUs but do see it's running on GPUs.  I gather it's slow because the GPU I have on my desktop is poor.

Maybe I'll kill it and run it on a CPU as that maybe faster on my local machine.

One thing that should be changed, sometime, is the only output I see is a warning message.

```
┌ Warning: ImmersedBoundaryGrid is unvalidated and may produce incorrect results. 
│ Don't hesitate to help validate ImmersedBoundaryGrid by reporting any bugs 
│ or unexpected behavior to https://github.com/CliMA/Oceananigans.jl/issues
└ @ Oceananigans.ImmersedBoundaries ~/software/Second_Oceananigans/Oceananigans.jl/src/ImmersedBoundaries/ImmersedBoundaries.jl:73
┌ Warning: ImmersedBoundaryGrid is unvalidated and may produce incorrect results. 
│ Don't hesitate to help validate ImmersedBoundaryGrid by reporting any bugs 
│ or unexpected behavior to https://github.com/CliMA/Oceananigans.jl/issues
└ @ Oceananigans.ImmersedBoundaries ~/software/Second_Oceananigans/Oceananigans.jl/src/ImmersedBoundaries/ImmersedBoundaries.jl:73
```

glwagner commented:
There's a progress writer:

https://github.com/CliMA/Oceananigans.jl/blob/8c60dba84334ddd1c731e1b5a3cc0414f9066cbd/validation/immersed_boundaries/internal_tide.jl#L28-L31

so the reason you didn't see output might be because it was still compiling??? Or, running very very very very slow.

francispoulin commented:
Agreed!  I went back and see that things are much better on my cpu. There are a few warnings but then it gets down to business.

I am estimating that it might take 14 hours, and now I believe that it is progressing (unlike before).

On a future world, when one sees that things are taking a while, it would be nice to run things in parallel using distrubed version of the model. I see there are distributed models for IncompressibleModel and ShallowWater, but not HydrostaticModel, yet.  

Maybe I'll create an issue since this is something that I presume we would like to have going forward.

```
┌ Warning: ImmersedBoundaryGrid is unvalidated and may produce incorrect results. 
│ Don't hesitate to help validate ImmersedBoundaryGrid by reporting any bugs 
│ or unexpected behavior to https://github.com/CliMA/Oceananigans.jl/issues
└ @ Oceananigans.ImmersedBoundaries ~/software/Second_Oceananigans/Oceananigans.jl/src/ImmersedBoundaries/ImmersedBoundaries.jl:73
┌ Warning: HydrostaticFreeSurfaceModel is experimental. Use with caution!
└ @ Oceananigans.Models.HydrostaticFreeSurfaceModels ~/software/Second_Oceananigans/Oceananigans.jl/src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl:106
┌ Warning: Cannot save Function property into grid/immersed_boundary/mask
└ @ Oceananigans.OutputWriters ~/software/Second_Oceananigans/Oceananigans.jl/src/OutputWriters/output_writer_utils.jl:21
┌ Warning: Could not initialize ./internal_tide.jld2: data may already be initialized.
└ @ Oceananigans.OutputWriters ~/software/Second_Oceananigans/Oceananigans.jl/src/OutputWriters/jld2_output_writer.jl:193
[ Info: [0.06%], iteration: 100, time: 0.055, max|w|: 9.46e-06
[ Info: [0.11%], iteration: 200, time: 0.110, max|w|: 2.67e-05
...
```

glwagner commented:
We would like a parallel hydrostatic model eventually, you're right. We need to parallelize the preconditioned conjugate gradient solver and ensure that reduced fields parallelize for that. I'm not sure this 2D internal tide problem is large enough to benefit from GPU parallelization --- have you tested GPU parallelization of such small problems?

francispoulin commented:
In #1722 we compared CPU vs GPU for the `ShallowWaterModel` for problems starting at 32x32 going as high as 16384x1634.  Not quite the same model miind you but for 256x256 we found the speed up to be 7.  This problem is 512x256 so it could be slightly beter but there is the pressure solve so this is just a guide.

I know that the Poisson solves are a bit part of it and agree this will take time.

glwagner commented:
How about speed up when going from 1 to 2 GPUs for 256^2?

francispoulin commented:
Last time @ali-ramadhan tried doing any benchmarking for 2 GPUs we found that the efficiency was around 50%, so it took as long for 1 GPU as it did for 2 GPU.  Henry is going to redo these benchmarks and I hope we will have some results by next week.

glwagner commented:
Ah ok, thanks for that info! I guess that means this internal tide validation experiment wouldn't benefit from GPU parallelization?

christophernhill commented:
> Last time @ali-ramadhan tried doing any benchmarking for 2 GPUs we found that the efficiency was around 50%, so it took as long for 1 GPU as it did for 2 GPU. Henry is going to redo these benchmarks and I hope we will have some results by next week.

@francispoulin @hennyg888 and @ali-ramadhan let me know if I can help with benchmarks. There are a bunch of silly things around CUDA enabled MPI, nvshmem that I have fiddled with in some other contexts.

francispoulin commented:
Thank you @christophernhill , that is kind of you to offer and I will certainly take you up on your offer.  

@hennyg888 has tried running things on 2 GPUs and has been getting an error at `Waitall!`, which is the first time that information is shared between GPUs.  I hope we can look into this a bit more and then ask you about it next week.

The last time that @ali-ramadhan successfully ran the distirbuted models on GPUs he found that the efficiency on 2 gpus was `50%`.  Not great. We are trying to reproduce this as a starting point and when we do, it'll be nice to figure out what's causing the bottleneck.  At the moment it's unknown to us but hope to start looking at this soon.

We will certainly ask you about this next week!

francispoulin commented:
> Ah ok, thanks for that info! I guess that means this internal tide validation experiment wouldn't benefit from GPU parallelization?

Not at this stage but hopefully sometime soon.
